### PR TITLE
Add Jiff date/time types `Zoned`/`TimeZone`

### LINKED ISSUE #1270: Provide support for Jiff date/time types
## Description

This is a proposal to add first-class support for the data types from [Jiff](https://crates.io/crates/jiff). First-class support is necessary, and an external crate would be much more inconvenient to use, because of Rust's orphan rule as outlined in https://github.com/graphql-rust/juniper/issues/87#issuecomment-325918618.

This should cover the following date/time types:

### Timestamps

- [`Timestamp`](https://docs.rs/jiff/0.1.4/jiff/struct.Timestamp.html), serialized as RFC 3339 (ISO 8601) date/time with UTC time zone, GraphQL scalar [`DateTime`](https://the-guild.dev/graphql/scalars/docs/scalars/date-time), e.g. `2024-06-19T19:22:45Z`
- ~[`Zoned`](https://docs.rs/jiff/0.1.4/jiff/struct.Zoned.html), serialized was RFC 3339 (ISO 8601) date/time with underlying time zone, GraphQL scalar `DateTime`, e.g. `2024-06-19T15:22:00-04:00`[^1]~

[^1]: ~Alternatively, this could include the time zone specifier and use RFC ~8536~ 9557 instead, i.e. `2024-07-04T08:39:00-04:00[America/New_York]`. This would probably be harder to consume, though, because that format seems not as widely used. It is also not a valid input for GraphQL scalar `DateTime`.~

### Wall-clocked

- [`Date`](https://docs.rs/jiff/0.1.4/jiff/civil/struct.Date.html), serialized as date, without time zone, GraphQL scalar [`LocalDate`](https://the-guild.dev/graphql/scalars/docs/scalars/local-date), e.g. `2024-06-19`
- [`Time`](https://docs.rs/jiff/0.1.4/jiff/civil/struct.Time.html), serialized as time, without time zone, GraphQL scalar [`LocalTime`](https://the-guild.dev/graphql/scalars/docs/scalars/local-time), e.g. `15:22:45`
- [`DateTime`](https://docs.rs/jiff/0.1.4/jiff/civil/struct.DateTime.html), serialized as date/time, without time zone, GraphQL scalar [`LocalDateTime`](https://the-guild.dev/graphql/scalars/docs/scalars/local-date-time), e.g. `2024-06-19T15:22:45`

### Durations

- [`Span`](https://docs.rs/jiff/0.1.4/jiff/struct.Span.html), serialized as ISO 8601 duration, GraphQL scalar [`Duration`](https://the-guild.dev/graphql/scalars/docs/scalars/duration), e.g. `P5dT8h1m`

The resulting serialization would be identical to the one for the [`chrono`](https://crates.io/crates/chrono) and [`time`](https://crates.io/crates/time) crates, with the difference of using GraphQL scalar `LocalDate` instead of [`Date`](https://the-guild.dev/graphql/scalars/docs/scalars/date).[^2]

[^2]: `LocalDate` seems to be the more appropriate type. I am not sure why `chrono` and `time` integration chose to use `Date` instead.

I am aware that Jiff is a very new library but I think it has great potential in the ecosystem. Having dealt with date/time intricacies in several projects, I regard it as one of the top-contenders for following or eventually even superseding `chrono`/`time`.

~I am willing to provide a PR if this proposal is accepted.~ See #1271 for a PR that implements these features.

### PR DESCRIPTION
**Pull request by sgoll**
## Description

This is a follow-up to #1271 and adds integration for the following [Jiff](https://crates.io/crates/jiff) date/time types:

- [`Zoned`](https://docs.rs/jiff/latest/jiff/struct.Zoned.html) to GraphQL scalar `ZonedDateTime`[^1]
- [`tz::TimeZone`](https://docs.rs/jiff/latest/jiff/tz/struct.TimeZone.html) to GraphQL scalar `TimeZoneOrUtcOffset`[^2], and to GraphQL scalar [`TimeZone`](https://the-guild.dev/graphql/scalars/docs/scalars/time-zone) with a newtype
- [`tz::Offset`](https://docs.rs/jiff/latest/jiff/tz/struct.Offset.html) to GraphQL scalar [`UtcOffset`](https://the-guild.dev/graphql/scalars/docs/scalars/utc-offset)

[^1]: This GraphQL scalar is not documented at https://graphql-scalars.dev but we can introduce it here as laid out in https://github.com/graphql-rust/juniper/issues/1270#issuecomment-2291041828.

[^2]: This GraphQL scalar does not exist yet either. It is the union of GraphQL scalars `TimeZone` and `UtcOffset`.

~The new types are hidden behind feature flag `jiff-tz` because they require enabling the `std` feature flag for the `jiff` crate (which is not necessary for the remaining Jiff date/time types introduced in #1271).~

In addition, the `jiff` crate itself must be installed either with the default feature flags or with explicit feature flags that make the Time Zone Database available as explained in [`jiff` time zone features](https://docs.rs/jiff/latest/jiff/index.html#time-zone-features).

Closes #1270

### CODE REVIEW

File: juniper/src/integrations/jiff.rs [hunk 1]
```diff
@@ -282,6 +334,51 @@ mod duration {
...
+    use super::*;
+
+    pub(super) fn to_output<S>(v: &TimeZone) -> Value<S>
+    where
+        S: ScalarValue,
+    {
+        Value::scalar(
+            v.iana_name()
+                .unwrap_or_else(|| panic!("Failed to format `TimeZone`: no IANA name"))
+                .to_owned(),
```
[the remark this answers is not in the document]
Comment (tyranron) [reply juniper/src/integrations/jiff.rs:368]: @sgoll non-fallible `to_output` is just the current limitation of the `#[graphql_scalar]`/`#[derive(GraphQLScalar)]` macros expansion and looks like this:
```rust
#[automatically_derived]
impl<__S> ::juniper::GraphQLValue<__S> for jiff::civil::DateTime
where
    __S: ::juniper::ScalarValue,
{
    type Context = ();
    type TypeInfo = ();
    fn type_name<'i>(&self, info: &'i Self::TypeInfo) -> ::core::option::Option<&'i ::core::primitive::str> { <Self as ::juniper::GraphQLType<__S>>::name(info) }
    fn resolve(&self, info: &(), selection: ::core::option::Option<&[::juniper::Selection<'_, __S>]>, executor: &::juniper::Executor<'_, '_, Self::Context, __S>) -> ::juniper::ExecutionResult<__S> {
        ::core::result::Result::Ok(local_date_time::to_output(self))
    }
```
In theory, it shouldn't b a very hard task to add there some `IntoResult` coercion. However, if we do this, the error will be returned to the client. And it's unclear to me whether it's better than just panicking, because failing to serialize a timezone value is not an expected error on the client, and more like something that shouldn't happen at all.

Usually, we solve this via new-type pattern by restricting certain values.

I see 2 options here:
1. Either we create `juniper::integrations::jiff::TimeZone` newtype over `jiff::TimeZone` and restrict constructing its values being non-IANA. This has ergonomics impact as provides a different type.
2. Or we just bother less, leave the `panic!` and mention in docs about the trick above (it's no deal for library users to define their own newtype and `derive(GraphQLScalar)` for it `transparent`ly if they need type safety for that case in their code).

File: juniper/src/integrations/jiff.rs [hunk 2]
```diff
@@ -2,26 +2,26 @@
...
-//! | [`civil::DateTime`] | `yyyy-MM-ddTHH:mm:ss` | [`LocalDateTime`][s3] |
-//! | [`Timestamp`]       | [RFC 3339] string     | [`DateTime`][s4]      |
-//! | [`Span`]            | [ISO 8601] duration   | [`Duration`][s5]      |
+//! | Rust type            | Format                | GraphQL scalar        |
+//! |----------------------|-----------------------|-----------------------|
+//! | [`civil::Date`]      | `yyyy-MM-dd`          | [`LocalDate`][s1]     |
+//! | [`civil::Time`]      | `HH:mm[:ss[.SSS]]`    | [`LocalTime`][s2]     |
+//! | [`civil::DateTime`]  | `yyyy-MM-ddTHH:mm:ss` | [`LocalDateTime`][s3] |
+//! | [`Timestamp`]        | [RFC 3339] string     | [`DateTime`][s4]      |
+//! | [`Zoned`][^1]        | [RFC 9557] string     | `ZonedDateTime`       |
```
[the remark this answers is not in the document]
Comment (tyranron) [reply juniper/src/integrations/jiff.rs:11]: @sgoll I cannot come up with better naming. I think it's OK. Once such scalar will appear, we always can rename and adjust it.

**Comment by tyranron** [comment]
@sgoll 

> The new types are hidden behind feature flag `jiff-tz` because they require enabling the `std` feature flag for the `jiff` crate (which is not necessary for the remaining Jiff date/time types introduced in #1271).

I think this is redundant. `juniper` at the moment is not capable of acting in `no_std`, so you could just add that feature and don't bother.

File: juniper/src/integrations/jiff.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BurntSushi) [reply juniper/src/integrations/jiff.rs:368]: Reiterating my ignorance of GraphQL here, using an offset if an IANA name isn't available is fine. It's what RFC 9557 recommends and what Jiff implements itself:

```rust
use jiff::{tz::TimeZone, Zoned};

fn main() -> anyhow::Result<()> {
    let zdt = Zoned::now();

    let tz = TimeZone::get("US/Eastern")?;
    println!("{}", zdt.with_time_zone(tz));

    let tz = jiff::tz::offset(-4).to_time_zone();
    println!("{}", zdt.with_time_zone(tz));

    Ok(())
}
```

Has this output:

```
$ cargo -q r
2024-08-15T20:50:50.12994762-04:00[US/Eastern]
2024-08-15T20:50:50.12994762-04:00[-04:00]
```

[the remark this answers is not in the document]
Comment (BurntSushi) [reply juniper/src/integrations/jiff.rs:368]: Can you just use the `Display` impl for `Zoned` for this?

If not, you can use `strftime("%:V")` to get a string that is "IANA time zone name OR offset":

```rust
use jiff::{tz::TimeZone, Zoned};

fn main() -> anyhow::Result<()> {
    let tz = TimeZone::get("US/Eastern")?;
    let zdt = Zoned::now().with_time_zone(tz);
    println!("{}", zdt);
    println!("{}", zdt.strftime("%:V"));

    let tz = jiff::tz::offset(-4).to_time_zone();
    let zdt = Zoned::now().with_time_zone(tz);
    println!("{}", zdt);
    println!("{}", zdt.strftime("%:V"));

    Ok(())
}
```

Has this output:

```
$ cargo -q r
2024-08-15T20:53:52.441756324-04:00[US/Eastern]
US/Eastern
2024-08-15T20:53:52.443774214-04:00[-04:00]
-04:00
```

Note that `%:V` (and `%V`) was introduced in `jiff 0.1.5`.

File: juniper/src/integrations/jiff.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (BurntSushi) [reply juniper/src/integrations/jiff.rs:11]: `ZonedDateTime` is indeed probably the right name for this. For example, [it's what Temporal uses](https://tc39.es/proposal-temporal/docs/zoneddatetime.html). It's also my second choice for Jiff, but I decided on a shorter name since "jiff" already provides the datetime context in the name. (And I prefer more concise names when it isn't too opaque.)

File: juniper/src/integrations/jiff.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tyranron) [reply juniper/src/integrations/jiff.rs]: > using an offset if an IANA name isn't available is fine. It's what RFC 9557 recommends

I would be happy to have it, unfortunately `TimeZone` GraphQL scalar implementation is [not capable of that](https://github.com/Urigo/graphql-scalars/blob/v1.23.0/src/scalars/TimeZone.ts). If we provide from the server side wider variety of values than the client could accept, it will break certain invariants and assumptions between them, maybe even silently.

If it only was about input coercion on the server side, this could have been OK, as we do for some `LocalDateTime` scalar implementations, being able to parse non-RFC-compliant `%Y-%m-%d %H:%i:%s` format from clients. However, we're talking here about returning values from server to client, and overcoming client expectations is not a good choice here, imho.

[the remark this answers is not in the document]
Comment (tyranron) [reply juniper/src/integrations/jiff.rs]: @sgoll 

> Where would be a good location to put the mention in the docs regarding `panic!()`?

I think module level docs are fine.

[the remark this answers is not in the document]
Comment (tyranron) [reply juniper/src/integrations/jiff.rs]: @sgoll 

> However, the same effort would be necessary to avoid panicking (if we reverted [ce43466](https://github.com/graphql-rust/juniper/commit/ce434667238f789ca0475222710c17e937427e38)). In both cases, a developer must avoid values that lie outside the strict interpretation of the `TimeZone` scalar,

Sorry, but I don't buy this argument. Because, while the server side is in developer's control here, the client may be not at all. So, having and controlling panics at the server side is much more convenient than having assumptions about clients doing that properly instead. Despite both are not perfect solutions.

> I just learned that there is a separate GraphQL scalar [`UTCOffset`](https://the-guild.dev/graphql/scalars/docs/scalars/utc-offset) which covers exactly those values we are after, after all.

Small remark: it's [named `UtcOffset`](https://github.com/Urigo/graphql-scalars/blob/f3359443959a485900bdaeec1a939ee575aa85a2/src/scalars/UtcOffset.ts#L21) after all. The naming on the site is a little bit fucked up there.

We do use it already for the `time::UtcOffset` type.


> So I propose the following solution:

I'm not against it. Removing subtle footguns is totally OK. I don't use `jiff` in my projects. For our purposes, just pure `time` is more than enough. I just hope those newtypes would not impose unconvenient ergonomics penalties for those who actually use it.

Another approach may be just naming the scalar like `TimeZoneOrUtcOffset` and say in the docs that it's either `TimeZone` or `UtcOffset` value. Actually, it's not another approach. We could provide both newtypes you've described, while making `jiff::tz::TimeZone` still a `TimeZoneOrUtcOffset` scalar.

> In theory, we _could_ also allow building `UTCOffset` scalar values from IANA time zone identifiers but we deliberately decided against it when not implementing [`DateTime`](https://the-guild.dev/graphql/scalars/docs/scalars/date-time) for `Zoned`.

I don't think `DateTime`/`Zoned` is relevant here. We do want to here to narrow the semantics of the `jiff::tz::TimeZone` type to semantics of the `UtcOffset` scalar. The result coercion in such cases is very common in GraphQL land. For example, the `LocalTime` scalar has milliseconds precision, while Rust type, of course, usually have micros and nanos. But we do the coercion to milliseconds when implementing this scalar.

So the situation where we have in the Rust code a `jiff::tz::TimeZone` defined as:
```rust
let tz = TimeZone::get("America/New_York")?;
```
But do want to return it to client as `UtcOffset`, because the client expects it like that, is quite normal and common, as for me, and we should not restrict it, giving the situation that we always may convert IANA to offset.

[the remark this answers is not in the document]
Comment (tyranron) [reply juniper/src/integrations/jiff.rs]: Maybe, we don't even need a second newtype:
- `jiff::tz::TimeZone` implementing `UtcOffset` scalar directly
- `juniper::integrations::jiff::TimeZone` newtype implementing `TimeZone` scalar

[the remark this answers is not in the document]
Comment (tyranron) [reply juniper/src/integrations/jiff.rs]: @sgoll 

> we cannot convert from IANA time zone identifier to UTC offset without also knowing the exact point in time that is relevant.

Ah, yes. My bet. Then let's stick to the solution with 3 scalars.

> The serialization to `UtcOffset` was an addition that became obvious when I worked on the other newtype—which is now not needed anymore.

Ah... ok! Nice! 👍

I'll try to review and merge it today.

File: book/src/types/scalars.md [hunk 3]
```diff
@@ -385,29 +385,35 @@ mod date_scalar {
...
+| [`chrono::DateTime`]        | [`DateTime`]          | [`chrono`]       |       |
+| [`chrono_tz::Tz`]           | [`TimeZone`]          | [`chrono-tz`]    |       |
+| [`rust_decimal::Decimal`]   | `Decimal`             | [`rust_decimal`] |       |
+| [`jiff::civil::Date`]       | [`LocalDate`]         | [`jiff`]         |       |
+| [`jiff::civil::Time`]       | [`LocalTime`]         | [`jiff`]         |       |
+| [`jiff::civil::DateTime`]   | [`LocalDateTime`]     | [`jiff`]         |       |
+| [`jiff::Timestamp`]         | [`DateTime`]          | [`jiff`]         |       |
+| [`jiff::Zoned`]             | `ZonedDateTime`       | [`jiff`]         |       |
+| [`jiff::tz::TimeZone`]      | `TimeZoneOrUtcOffset` | [`jiff`]         |       |
+| [`jiff::tz::TimeZone`]      | [`TimeZone`]          | [`jiff`]         | [^n1] |
```
Comment (LegNeato) [inline book/src/types/scalars.md:405]: Not sure we want a whole column just for this, put this on the scalar?

File: book/src/types/scalars.md [hunk 4]
```diff
@@ -385,29 +385,35 @@ mod date_scalar {
...
+| [`jiff::Span`]              | [`Duration`]          | [`jiff`]         |       |
+| [`time::Date`]              | [`LocalDate`]         | [`time`]         |       |
+| [`time::Time`]              | [`LocalTime`]         | [`time`]         |       |
+| [`time::PrimitiveDateTime`] | [`LocalDateTime`]     | [`time`]         |       |
+| [`time::OffsetDateTime`]    | [`DateTime`]          | [`time`]         |       |
+| [`time::UtcOffset`]         | [`UtcOffset`]         | [`time`]         |       |
+| [`url::Url`]                | [`URL`]               | [`url`]          |       |
+| [`uuid::Uuid`]              | [`UUID`]              | [`uuid`]         |       |
+
+[^n1]: Conversion supported via newtype.
```
Comment (LegNeato) [inline book/src/types/scalars.md:416]: Would be nice to link to docs here.

File: book/src/types/scalars.md [hunk 5]
```diff
@@ -424,6 +430,7 @@ mod date_scalar {
 [`chrono::NaiveTime`]: https://docs.rs/chrono/latest/chrono/naive/struct.NaiveTime.html
 [`chrono-tz`]: https://docs.rs/chrono-tz
 [`chrono_tz::Tz`]: https://docs.rs/chrono-tz/latest/chrono_tz/enum.Tz.html
+[`rust_decimal::Decimal`]: https://docs.rs/rust_decimal/latest/rust_decimal/struct.Decimal.html
```
Comment (LegNeato) [inline book/src/types/scalars.md:433]: Is looks unrelated and there is already `Decimal` below

File: juniper/src/integrations/jiff.rs [hunk 6]
```diff
@@ -282,6 +348,231 @@ mod duration {
...
+        match self {
+            Self::InvalidTimeZone(err) => Some(err),
+            Self::MissingIanaName(_) => None,
+        }
+    }
+}
+
+/// Representation of time zone.
+///
+/// Is a set of rules for determining the civil time, via an offset from UTC, in a particular
```
Comment (LegNeato) [inline juniper/src/integrations/jiff.rs:443]:
```suggestion
/// A set of rules for determining the civil time, via an offset from UTC, in a particular
```

**Review by LegNeato — CHANGES_REQUESTED** [review]
Some minor nits. I'll give @tyranron a chance to look over but LGTM overall! Thanks for putting up such high quality PRs.

**Comment by LegNeato** [comment]
Failures Re unrelated. Thanks again!
